packages/plugins/rows/src/editor/render.tsx

```diff
@@ -272,26 +323,76 @@ export function RowRenderer({
     }
   }, [drag, store, dispatch, index, row.id, rows])
 
-  dragPreview(drop(container))
+  dragPreview(drop(dropContainer))
+  const focusPath = useScopedSelector(getFocusPath(row.id))
+  const focusTree = useScopedSelector(getFocusTree())
+
+  const dropPreview =
+    canDrop(collectedDropProps.id) && collectedDropProps.id ? (
+      <Inserted>
+        <SubDocument id={collectedDropProps.id} />
+      </Inserted>
+    ) : null
 
   return (
-    <div ref={container}>
-      <div>{row.render(pluginProps)}</div>
-    </div>
+    <React.Fragment>
+      {draggingAbove ? dropPreview : null}
+      <div ref={container}>
+        {collectedDragProps.isDragging ? (
+          <GrayOut>{row.render(pluginProps)}</GrayOut>
+        ) : (
+          <div>{row.render(pluginProps)}</div>
+        )}
+      </div>
+      {!draggingAbove ? dropPreview : null}
+    </React.Fragment>
   )
-}
 
-function isDraggingAbove(monitor: DropTargetMonitor) {
-  // get bounding Rect set in hover
-  const domBoundingRect = monitor.getItem().boundingRect
+  function isDraggingAbove(monitor: DropTargetMonitor) {
+    if (!container.current) {
+      return false
+    }
+    const domBoundingRect = container.current.getBoundingClientRect()
+
+    const domMiddleY = (domBoundingRect.bottom - domBoundingRect.top) / 2
+    const dropClientOffset = monitor.getClientOffset()
+    const dragClientY = dropClientOffset
+      ? dropClientOffset.y - domBoundingRect.top
+      : 0
+
+    return dragClientY < domMiddleY
+  }
+
+  function canDrop(draggedId: string | null) {
+    if (!collectedDropProps.isDragging || !draggedId) {
+      return false
+    }
+    if (!focusPath || focusPath.includes(draggedId)) {
+      // dropzone is child of dragged element or element itself
+      return false
+    }
 
-  const domMiddleY = (domBoundingRect.bottom - domBoundingRect.top) / 2
-  const dropClientOffset = monitor.getClientOffset()
-  const dragClientY = dropClientOffset
-    ? dropClientOffset.y - domBoundingRect.top
-    : 0
+    if (!focusTree) return false
+    const parent = findParent(focusTree, draggedId)
+    if (!parent || !parent.children) return false
 
-  return dragClientY < domMiddleY
+    const dragIndex = R.findIndex(
+      node => node.id === draggedId,
+      parent.children
+    )
+    const dropZoneIndex = R.findIndex(
+      node => node.id === row.id,
+      parent.children
+    )
+    if (dragIndex === -1 || dropZoneIndex === -1) {
+      // different parent
+      return false
+    }
+    // check that its not the same position
+    return draggingAbove
+      ? dropZoneIndex - 1 !== dragIndex
+      : dropZoneIndex + 1 !== dragIndex
+  }
```

Comment:
@inyono any ideas of refactoring this to make it clearer what's happening?

Reply:
Done